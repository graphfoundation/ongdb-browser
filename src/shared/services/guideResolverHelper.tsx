/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import MdxSlide from 'browser/modules/Docs/MDX/MdxSlide'
import Slide from 'browser/modules/Carousel/Slide'
import docs, { isGuideChapter } from 'browser/documentation'
import guideUnfound from 'browser/documentation/sidebar-guides/unfound'
import {
  addProtocolsToUrlList,
  extractAllowlistFromConfigString,
  resolveAllowlistWildcard
} from './utils'
import { fetchRemoteGuide } from 'shared/modules/commands/helpers/play'
import {
  getDefaultRemoteContentHostnameAllowlist,
  getRemoteContentHostnameAllowlist
} from 'shared/modules/dbMeta/dbMetaDuck'
import { splitMdxSlides } from 'browser/modules/Docs/MDX/splitMdx'
import {
  StyledCypherErrorMessage,
  StyledDiv,
  StyledErrorH4,
  StyledHelpContent,
  StyledHelpDescription,
  StyledHelpFrame,
  StyledPreformattedArea
} from 'browser/modules/Stream/styled'

const { chapters } = docs.guide

export async function resolveGuide(
  guideName: string,
  state: any
): Promise<{ slides: JSX.Element[]; title: string }> {
  const isUrl = guideName.startsWith('http')
  if (isUrl) {
    return await resolveRemoteGuideFromURL(guideName, state)
  }

  if (isGuideChapter(guideName)) {
    return chapters[guideName]
  }

  try {
    return await resolveRemoteGuideFromName(guideName, state)
  } catch (e) {}

  return guideUnfound
}

function mdxTextToSlides(mdx: string): JSX.Element[] {
  return splitMdxSlides(mdx).map((slide, index) => (
    // index is fine since we'll never move or delete slides
    <MdxSlide key={index} mdx={slide} isSidebarSlide />
  ))
}

function htmlTextToSlides(html: string): JSX.Element[] {
  const tmpDiv = document.createElement('div')
  tmpDiv.innerHTML = html
  const htmlSlides = tmpDiv.getElementsByTagName('slide')
  if (htmlSlides && htmlSlides.length) {
    return Array.from(htmlSlides).map((slide, index) => (
      <Slide key={index} html={slide.innerHTML} isSidebarSlide />
    ))
  }
  return [<Slide key="first" html={html} isSidebarSlide />]
}

async function resolveRemoteGuideFromURL(
  guideName: string,
  state: any
): Promise<{ slides: JSX.Element[]; title: string }> {
  const url = guideName
  const urlObject = new URL(url)
  urlObject.href = url
  const filenameExtension =
    (urlObject.pathname.includes('.') && urlObject.pathname.split('.').pop()) ||
    'html'
  const allowlist = getRemoteContentHostnameAllowlist(state)

  try {
    const remoteGuide = await fetchRemoteGuide(url, allowlist)
    const titleRegexMatch = remoteGuide.match(/<title>(.*?)<\/title>/)
    const title = (titleRegexMatch && titleRegexMatch[1])?.trim() || guideName
    if (['md', 'mdx'].includes(filenameExtension)) {
      return {
        slides: mdxTextToSlides(remoteGuide),
        title
      }
    } else {
      return {
        slides: htmlTextToSlides(remoteGuide),
        title
      }
    }
  } catch (e) {
    if (e.response && e.response.status === 404) {
      return guideUnfound
    }
    return {
      title: 'Error',
      slides: [
        <Slide key="first" isSidebarSlide>
          <StyledHelpFrame>
            <StyledHelpContent>
              <StyledHelpDescription>
                <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
                <StyledErrorH4>Remote guide error</StyledErrorH4>
              </StyledHelpDescription>
              <StyledDiv>
                <StyledPreformattedArea>
                  {e.name}: {e.message}
                </StyledPreformattedArea>
              </StyledDiv>
            </StyledHelpContent>
          </StyledHelpFrame>
        </Slide>
      ]
    }
  }
}

async function resolveRemoteGuideFromName(
  guideName: string,
  state: any
): Promise<{ slides: JSX.Element[]; title: string }> {
  const allowlistStr = getRemoteContentHostnameAllowlist(state)
  const allowlist = extractAllowlistFromConfigString(allowlistStr)
  const defaultAllowlist = extractAllowlistFromConfigString(
    getDefaultRemoteContentHostnameAllowlist()
  )
  const resolvedWildcardAllowlist = resolveAllowlistWildcard(
    allowlist,
    defaultAllowlist
  )
  const urlAllowlist = addProtocolsToUrlList(resolvedWildcardAllowlist)
  const possibleGuidesUrls: string[] = urlAllowlist.map(
    (url: string) => `${url}/${guideName}`
  )

  return possibleGuidesUrls
    .reduce(
      (promiseChain: Promise<any>, currentUrl: string) =>
        promiseChain
          .catch(() => fetchRemoteGuide(currentUrl, allowlistStr))
          .then(r => Promise.resolve(r)),
      Promise.reject(new Error())
    )
    .then(text => ({
      slides: htmlTextToSlides(text),
      title: guideName
    }))
}
