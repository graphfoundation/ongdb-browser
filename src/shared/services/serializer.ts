/*
 * Copyright (c) "Graph Foundation"
 * Graph Foundation, Inc. [https://graphfoundation.org]
 *
 * This file is part of ONgDB.
 *
 * ONgDB is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

const csvDelimiter = ','
const csvNewline = '\n'

const csvEscape = (str: any) => {
  if (!isString(str)) return str
  if (isEmptyString(str)) return '""'
  if (hasQuotes(str) || hasDelimiterChars(str) || hasNewLines(str)) {
    return `"${str.replace(/"/g, '""')}"`
  }
  return str
}
const serializeObject = (input: any) =>
  isObject(input) ? JSON.stringify(input) : input

const hasDelimiterChars = (str: any) => str && str.indexOf(csvDelimiter) > -1
const hasNewLines = (str: any) => str && str.indexOf(csvNewline) > -1
const hasQuotes = (str: any) => str && str.indexOf('"') > -1
const isString = (str: any) => typeof str === 'string'
const isObject = (str: any) => typeof str === 'object' && str !== null
const isEmpty = (str: any) => typeof str === 'undefined' || str === null
const isEmptyString = (str: any) => str === ''

const csvChain = (input: any) =>
  (Array.isArray(input) ? input : [])
    .map(serializeObject)
    .map(csvEscape)
    .join(csvDelimiter)

export const CSVSerializer = (cols: any) => {
  const _cols = cols
  const _data: any = []
  const append = (row?: any) => {
    const emptyRowInOneCol = isEmpty(row) && _cols.length === 1
    if (emptyRowInOneCol) return _data.push(row)
    if (!row || row.length !== _cols.length) {
      throw new Error('Column number mismatch')
    }
    _data.push(row)
  }
  return {
    append,
    appendRows: (rows: any) => rows.forEach(append),
    output: () =>
      csvChain(_cols) +
      (!_data.length ? '' : csvNewline + _data.map(csvChain).join(csvNewline))
  }
}
