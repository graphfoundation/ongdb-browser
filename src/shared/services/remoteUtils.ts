/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* global btoa */
import { parseURLWithDefaultProtocol } from 'services/utils'

const removeJavascriptFromHref = (string: any) => {
  const localString = string.replace(/href=".*javascript:[^"]*"/, 'href=""')
  return localString.replace(/href='.*javascript:[^']*'/, "href=''")
}
const removeScriptTags = (string: any) =>
  string.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*(<\/script>)?/gi, '')
const removeOnHandlersFromHtml = (string: any) =>
  string.replace(
    /(\s+(on[^\s=]+)[^\s=]*\s*=\s*("[^"]*"|'[^']*'|[\w\-.:]+\s*))/gi,
    ''
  )

export function cleanHtml(string: string): string {
  if (typeof string !== 'string') return string
  const stringWithoutHandlers = removeOnHandlersFromHtml(string)
  const stringWithoutScript = removeScriptTags(stringWithoutHandlers)
  return removeJavascriptFromHref(stringWithoutScript)
}

export const authHeaderFromCredentials = (username: any, password: any) => {
  if (!btoa) throw new Error('btoa not defined') // Non browser env
  return btoa(`${username}:${password}`)
}

export const isLocalRequest = (
  localUrl?: string | null,
  requestUrl?: string | null,
  opts = { hostnameOnly: false }
): boolean => {
  if (!localUrl || !requestUrl) return false

  const localUrlInfo = parseURLWithDefaultProtocol(localUrl)
  if (!localUrlInfo) return false

  if (requestUrl.startsWith('/')) return true

  const requestUrlInfo = parseURLWithDefaultProtocol(requestUrl)
  if (!requestUrlInfo) return false

  if (!requestUrlInfo.host) return true // GET /path

  if (opts.hostnameOnly === true) {
    return requestUrlInfo.hostname === localUrlInfo.hostname
  } // GET localhost:8080 from localhost:9000

  if (
    requestUrlInfo.host === localUrlInfo.host &&
    requestUrlInfo.protocol === localUrlInfo.protocol
  ) {
    // Same host and protocol
    return true
  }
  return false
}
