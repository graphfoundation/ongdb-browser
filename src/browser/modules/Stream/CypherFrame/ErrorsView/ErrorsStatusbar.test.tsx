/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { render } from '@testing-library/react'
import React from 'react'

import { ErrorsStatusbar } from './ErrorsStatusbar'
import { BrowserError } from 'services/exceptions'

describe('ErrorsStatusbar', () => {
  test('displays nothing if no error', () => {
    // Given
    const props = {
      result: null
    }

    // When
    const { container } = render(<ErrorsStatusbar {...props} />)
    expect(container).toMatchSnapshot()
  })
  test('displays error', () => {
    // Given
    const error: BrowserError = {
      code: 'Test.Error',
      message: 'Test error description',
      type: 'Neo4jError'
    }
    const props = {
      result: error
    }

    // When
    const { container } = render(<ErrorsStatusbar {...props} />)
    expect(container).toMatchSnapshot()
  })
})
