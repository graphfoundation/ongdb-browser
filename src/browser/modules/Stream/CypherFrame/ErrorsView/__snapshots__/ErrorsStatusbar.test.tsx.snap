// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorsStatusbar displays error 1`] = `
<div>
  <div
    class="sc-fFeiMQ kjEgmO"
  >
    <span
      class="sc-jeraig hOFtwq"
      title="Test.Error: Test error description"
    >
      <i
        class="sc-hBUSln jKTDNL fa fa-exclamation-triangle"
        style="line-height: inherit;"
      />
       
      Test.Error: Test error description
    </span>
  </div>
</div>
`;

exports[`ErrorsStatusbar displays nothing if no error 1`] = `<div />`;
